### ISSUE TITLE
9FUS Y31 firmware 8.2.0

### ISSUE THREAD

**Issue by scorpio862** [roleoroleo/yi-hack-MStar#91]
I just bought a YI Home1080p 9FUS Y31 with firmware 8.2.0. Is there any chance of getting this hack to work with this new firmware?

**Comment by roleoroleo** [comment]
To check if the hack works, I need a dump of your firmware.
Please read the wiki:
https://github.com/roleoroleo/yi-hack-MStar/wiki/Dump-your-backup-firmware

**Comment by ruppert92** [comment]
Is there anyway to rollback firmware? I also have version 8.2, but only after accidentally updating in the app

**Comment by roleoroleo** [comment]
Do you remember the exact string of the old version?
Like 4.5.0.0C_201910080934

**Comment by ruppert92** [comment]
Unfortunately no I never saw the original number

**Comment by roleoroleo** [comment]
No way.

**Comment by micfogas** [comment]
I have this cam too. 8.2.0.0 firmware. I didn't think to hack it when I first set it up. I did a firmware update on it pretty much instantly. the previous firmware version was also 8.2.0.0 but an earlier date. I lost my USB serial cable so I'm gonna grab another off Amazon and dump the firmware for you. nothing happens so far with 2 SD cards I've tried. I even tried getting into the manual update mode to cap the firmware file but it's stored in the root-only data folder on my phone. hate to reroot just for that but I can easily. tried packet capping using laptop as hotspot for the cam, just to get the firmware file names at least. I'm guessing they renamed them maybe for the latest. I'm gonna keep working at it. I'll update you.

**Comment by WattsC-90** [comment]
going to be following this thread keenly.. I have also run into this issue it seems! 8.2.0.0 firmware on a 9FUS 1080p home cam. how dare they change things! :p

**Comment by starbuck93** [comment]
What should the log output look like? @roleoroleo Here's my first attempt at getting the logs.
![image](https://user-images.githubusercontent.com/5626027/75613625-26078f00-5aed-11ea-9a1b-f9e63430054a.png)
![image](https://user-images.githubusercontent.com/5626027/75613631-43d4f400-5aed-11ea-8b7b-c63069fcd36f.png)

**Comment by roleoroleo** [comment]
It seems very very different from the platform I know.
Please try to repeat with SD inserted.

**Comment by starbuck93** [comment]
printenv outputs:
https://hastebin.com/newejicuce.makefile <-- before I inserted an SD card
https://hastebin.com/lutedolugo.makefile <-- after I inserted an SD card

full raw log output with SD card --> https://hastebin.com/ifikotuxod.sql

**Comment by WattsC-90** [comment]
@starbuck93 line 171, `---update file(home_y20gam) Not exist---` ...
@roleoroleo does this mean wrong file name with new version?

**Comment by roleoroleo** [comment]
Unfortunately it means unsupported platform.

**Comment by starbuck93** [comment]
Is there any other info I can grab off of the device for you?

**Comment by starbuck93** [comment]
@WattsC-90 yeah I didn't have the update files on the SD card to attempt to flash it. I tried pretty much all of the firmware hacks before I ended up soldering these tiny wires to the circuit board to get this data.

**Comment by roleoroleo** [comment]
It's a new platform.
Incompatible with v4 and Mstar.
I think we need a new hack.

**Comment by roleoroleo** [comment]
@starbuck93 
Are you able to make a dump as explained in my wiki?
If yes, we need to modify a bit the script and the flash addresses.

**Comment by starbuck93** [comment]
@roleoroleo I should be able to. I attempted to yesterday but only had a few minutes available and didn't know how to figure out the length of the parameters of the script.

**Comment by WattsC-90** [comment]
@roleoroleo how easy is it to create a new hack?

**Comment by roleoroleo** [comment]
@starbuck93 
The log doesn't report the mtd partition.
We have only this:
`partitions=boot@mtdblock1:rootfs@mtdblock2:home@mtdblock3:env@mtdblock4:mfg@mtdblock5:conf@mtdblock6:UDISK@mtdblock7`
Probably we need to dump the whole flash... very long operation.

@WattsC-90 
It depends on how many differences there are. And I don't know if I can help because I don't have this model.

**Comment by starbuck93** [comment]
@roleoroleo how would I go about that? Do you think using the python script to grab every byteword from 0 to 64Mb or however large? I found this command `md` and it outputs this: https://hastebin.com/xagilehete.makefile

**Comment by roleoroleo** [comment]
Yes, the python script runs the md command...
The flash is a en25qh128, so 16 MB.

**Comment by starbuck93** [comment]
I ran `md 0 8000000` before I saw your previous message. It just stopped so let me show you what it outputted. https://drive.google.com/open?id=1-r8_MsrR2rx3vHyKjEDHSGUTghIczIjM

**Comment by WattsC-90** [comment]
Any luck on this one? unfortunately I am entirely unable to help piece things together for you or get anything together :(

**Comment by alpacalife** [comment]
Just bought one of these cameras! I never upgraded the firmware and out of the box it was version 8.2.0.0A_201912121707

I can attempt to get a dump if @starbuck93 doesn't have any luck. I need to review the wiki a bit.

**Comment by roleoroleo** [comment]
You have to change the script.
The following line:
`if line == 'MStar # ':`
must be changed according to your u-boot prompt.
Pay attention to the space after the '#'.

**Comment by rclsilver** [comment]
Hello, i've this version too (4.5.0.0C_201910080934), how can I help you?

I tried to put file on SD but camera boots without flash... I can, if required, dump my rom

**Comment by roleoroleo** [comment]
4.5.0.0C is probably already supported.
See y203c files.

**Comment by rclsilver** [comment]
Hum, you're right... I'll try this evening... thanks!

**Comment by rogerb831** [comment]
I just bought a BFUS camera with version 8.2.0.0A_201912121707. It also wants me to do an upgrade to 8.2.0.0A_201912270941

**Comment by rogerb831** [comment]
I just ordered a couple more for sacrificial use and a USB/TTL adapter. I'll try to dump if nobody beats me to it.

**Comment by rogerb831** [comment]
Ok, got connected and grabbed a boot log. https://www.dropbox.com/s/pbvjdh3omhnocwk/bootlog.txt?dl=0. It's nearly identical to Starbuck's.  I tried to follow the directions to dump the flash, but theres no indicator of where the partitions are. Then I attempted to just dump 0 to 16MB using the python script, but the platform doesn't seem to include the "sf" command. Any tips? Maybe I'm missing something?

**Comment by roleoroleo** [comment]
Yes.
I think it's the same situation as @starbuck93.
Did you stop correctly the boot at the bootloader?

**Comment by rogerb831** [comment]
Sorry, I;m new to embedded and didn't see that in the wiki page. I was able to locate it. FYI, for this platform the string to watch for is 
`"Hit any key to stop autoboot: "`

sf probe gives the following
`=> sf probe
sf: bus-0 cs-0 spedd-50000000Hz mode-3000
SF: Detected en25qh128 with page size 256 Bytes, erase size 64 KiB, total 16 MiB`

So I can use the python script by updating the "MStar # " string and giving it the hex range 0x00 to 0xf42400 to  just grab the whole 16MB?

**Comment by roleoroleo** [comment]
You need to dump the whole flash:
```
sf probe
sf read 0x21000000 0x0 0x1000000
py dump.py --timeout 1 --step 64 comX 115200 0x21000000 0xfffff0
```
A lot of patience.

**Comment by alpacalife** [comment]
Just to give an update. I got mine connected over serial. I am able to connect and work with it.

I saw the same issue @starbuck93 saw, not mtd partitions defined. Don't know what to dump.

I got to the point of wanting to dump the entire 16MB flash, but I got an error or something when using sf command to dump to memory... I then gave up and went to bed. 

@roleoroleo I can test those commands tonight and post results unless I have any errors. those look like what I recall trying to do!

**Comment by Nanook** [comment]
I have the same Cam. New to this stuff. Happy to crack it open to confirm anything. Came with 8.2.0.0A_201912270941 installed.

**Comment by rogerb831** [comment]
Thanks roleoroleo, that helped. The dump is running now. This is going to take some time. Slight modification to your commands for the others who are playing along.

`sf probe
sf read 0x21000000 0x0 0x1000000
py dump.py --timeout 1 --step 64 comX 115200 0x21000000 0x22000000`

And here is the modified script that works for this platform

`#!/usr/bin/env python3
__author__ = 'Ion Agorria'
import argparse
from datetime import datetime
from time import time
from serial import Serial


def visible(text):
    text.replace("\n", "\\n")
    text.replace("\r", "\\r")
    return text


def parse_line(line, size):
    # Break down each part
    if len(line) != size:
        raise Exception("Line size %s doesn't match %s: %s" %
                        (len(line), size, visible(line)))
    i = 0
    addr = line[i:i + 8]
    i += 8
    assert line[i:i + 2] == ": "
    i += 2
    hex_data = line[i:i + 8]
    i += 8
    assert line[i:i + 1] == " "
    i += 1
    hex_data += line[i:i + 8]
    i += 8
    assert line[i:i + 1] == " "
    i += 1
    hex_data += line[i:i + 8]
    i += 8
    assert line[i:i + 1] == " "
    i += 1
    hex_data += line[i:i + 8]
    i += 8
    assert line[i:i + 4] == "    "
    i += 4
    text = line[i:i + 16]
    i += 16
    assert line[i:i + 2] == "\r\n"

    # Convert hex string into actual data
    data = []
    for i in range(0, len(hex_data), 2):
        piece = hex_data[i:i + 2]
        data.append(int(piece, 16))

    # For redundancy check if data matches text, but only for printable characters (non dot)
#    for i, piece in enumerate(data):
#        if text[i] != ".":
#            assert chr(piece) == text[i]

    return int(addr, 16), data, text


def write(serial, opts, data):
    if opts.debug:
        print("Debug write: " + visible(data))
    data = bytes(data, "ascii")
    serial.write(data)


def dump(serial, opts, log):
    data = []
    finish = False
    start_addr = opts.start
    last_addr = start_addr - 0x10
    do_reset = False

    # Restore the previous dump
    if opts.previous:
        loaded = False
        last_percentage = 0
        print("Info: loading from " + opts.previous)
        with open(opts.previous, 'r') as previous:
            lines = previous.readlines()
            count = len(lines)
            i = 0
            for line in lines:
                # Fix lines with only \n
                if line[-2:] != "\r\n":
                    line = line[:-1] + "\r\n"

                addr, line_data, text = parse_line(line, opts.size)

                # Check if we skipped some line
                if last_addr != addr - 0x10:
                    raise Exception("Possible skip or repetition, last address 0x%x doesn't match with previous address 0x%x" % (
                        last_addr, addr - 0x10))
                last_addr = addr

                # Check if we reach end
                if addr > opts.end:
                    print("Info: Reached specified end address")
                    finish = True
                    break

                # Discard if not start
                if addr < opts.start:
                    print(
                        "Warning: address %s is lower than start address! discarding" % addr)
                    continue

                # Store line data and log
                data.append(bytes(line_data))
                if not opts.ignore_log:
                    log.write(line)
                    log.flush()

                # Print percentage
                i += 1
                percentage = int(round(i / count * 100))
                if last_percentage != percentage:
                    print("Info: (%i%%) %i/%i " % (percentage, count, i))
                last_percentage = percentage

            # If we reach here everything went fine
            loaded = True

        if not loaded:
            raise Exception(
                "Dump was not restored completely, something went wrong")

        start_addr = last_addr + 0x10
        last_addr = start_addr - 0x10
        print("Info: finished loading previous")

    if not finish:
        # Send the initial command
        write(serial, opts, "md %s %s\n" %
              (hex(start_addr), hex(opts.step * 4)))
        do_reset = opts.reset

    while not finish:
        # Read response
        chunk = serial.readlines()
        if opts.debug:
            print("Debug read: " + str(chunk))

        assert len(chunk) > 0

        # Remove first one, its the command that we sent
        chunk = chunk[1:]

        # Iterate each line in chunk
        for line in chunk:
            line = line.decode("ascii")

            if opts.debug:
                print("line: " + str(line))

            # Adquire new chunk by sending newline
            if line == '=> ':
                if opts.debug:
                    print("Debug: detected prompt, sending newline")
                write(serial, opts, "\n")
                continue

            addr, line_data, text = parse_line(line, opts.size)

            # Check if we skipped some line
            if last_addr != addr - 0x10:
                raise Exception("Possible skip or repetition, last address 0x%x doesn't match with previous address 0x%x" % (
                    last_addr, addr - 0x10))
            last_addr = addr

            # Check if we reach end
            if addr > opts.end:
                print("Info: Reached specified end address")
                finish = True
                break

            # Discard if not start
            if addr < opts.start:
                print(
                    "Warning: address %s is lower than start address! discarding" % addr)
                continue

            # Print current line
            hex_addr = hex(addr).upper()[2:]
            while not len(hex_addr) == 8:
                hex_addr = "0" + hex_addr
            hex_data = ""
            for line_byte in line_data:
                line_byte = hex(line_byte).upper()[2:]
                if len(line_byte) == 1:
                    line_byte = "0" + line_byte
                hex_data += " " + line_byte
            print("0x%s %s |%s|" % (hex_addr, hex_data, text))

            # Store line data and log
            data.append(bytes(line_data))
            log.write(line)
            log.flush()

        if do_reset:
            print("Info: sending reset as requested")
            write(serial, opts, "reset\n")

    return data


def main():
    # Args parse
    parser = argparse.ArgumentParser()
    parser.add_argument("port", help="Serial port of device")
    parser.add_argument("baud", type=int, help="Serial baud rate")
    parser.add_argument(
        "start",  help="Start address in dec or hex (with 0x), must be multiple of 16")
    parser.add_argument(
        "end", help="End address in dec or hex (with 0x), must be multiple of 16")
    parser.add_argument("--step", type=int, default=256,
                        help="Number of lines per dump chunk")
    parser.add_argument("--size", type=int, default=67,
                        help="Total size of each line including spaces and newlines")
    parser.add_argument("--timeout", type=float, default=0.1,
                        help="Timeout in secs for serial")
    parser.add_argument("--previous", help="Previous log to continue from")
    parser.add_argument("--ignore-log", action='store_true',
                        help="Previous log is not saved on new log")
    parser.add_argument('--debug', action='store_true',
                        help='Enables debug mode')
    parser.add_argument('--reset', action='store_true',
                        help='Sends "reset" after finishing')
    opts = parser.parse_args()

    # Args conversion
    if opts.start[0:2] == "0x":
        opts.start = int(opts.start, 16)
    else:
        opts.start = int(opts.start)
    if opts.end[0:2] == "0x":
        opts.end = int(opts.end, 16)
    else:
        opts.end = int(opts.end)

    print(opts)

    # Args check
    if opts.start % 16 != 0:
        raise Exception("start argument is not multiple of 16")
    if opts.start < 0:
        raise Exception("start argument is too low")
    if opts.end % 16 != 0:
        raise Exception("end argument is not multiple of 16")
    if opts.end <= opts.start:
        raise Exception("end argument is too low")
    if opts.step <= 0:
        raise Exception("step argument is too low")
    if opts.size <= 0:
        raise Exception("size argument is too low")
    if opts.timeout <= 0:
        raise Exception("timeout argument is too low")
    if opts.debug:
        print("Debug mode enabled")
    name = datetime.fromtimestamp(time()).strftime(
        "%Y-%m-%dT%H%M%S") + "_" + hex(opts.start) + "_" + hex(opts.end)

    # Prepare to dump
    serial = Serial(port=opts.port, baudrate=opts.baud, timeout=opts.timeout)
    log = open(name + ".log", "w")
    try:
        data = dump(serial, opts, log)
    except Exception as e:
        raise e
    finally:
        serial.close()
        log.close()

    # Write to file
    file = open(name + ".img", "wb")
    try:
        for line in data:
            file.write(line)
    except Exception as e:
        raise e
    finally:
        file.close()


main()
`

**Comment by rogerb831** [comment]
I just realized I was getting all zeroes. Turns out the sf probe failed. I'm a bit stuck on this one. I tried using different start addresses but get the same error.

```
=> sf read 0x21000000 0x0 0x1000000
device 0 whole chip
rx wait_dma_recv_over fail
SF: 16777216 bytes @ 0x0 time:5005 ms Read: ERROR -1
```

**Comment by rogerb831** [comment]
ok, did some more exploration and found that the uboot command "fdt print /" lists out the device tree. In there I found the following.
```

                partitions {
                        device_type = "partitions";
                        boot {
                                device_type = "boot";
                                offset = <0x00000020>;
                                size = <0x00001200>;
                        };
                        rootfs {
                                device_type = "rootfs";
                                offset = <0x00001220>;
                                size = <0x00000c00>;
                        };
                        home {
                                device_type = "home";
                                offset = <0x00001e20>;
                                size = <0x00005c00>;
                        };
                        env {
                                device_type = "env";
                                offset = <0x00007a20>;
                                size = <0x00000100>;
                        };
                        mfg {
                                device_type = "mfg";
                                offset = <0x00007b20>;
                                size = <0x00000080>;
                        };
                        conf {
                                device_type = "conf";
                                offset = <0x00007ba0>;
                                size = <0x00000080>;
                        };
                        UDISK {
                                device_type = "UDISK";
                                offset = <0x00007c20>;
                                size = <0x00000000>;
                        };
                };

```

**Comment by roleoroleo** [comment]
Try to use sunxi_flash command.
You can have an example looking at the printenv output.

I think this is the help:
```
Usage:
sunxi_flash read command parmeters :
parmeters 0 : addr to load(hex only)
parmeters 1 : the name of the part to be load
[parmeters 2] : the number of bytes to be load(hex only)
if [parmeters 2] not exist, the number of bytes to be load is the size of the part indecated on partemeter 1
```

**Comment by rogerb831** [comment]
Ok, I made some progress on this. sunxi_flash seems to be the way to do it. It wants you to give it a RAM address to load into to and a partition name to read from the flash

`sunxi_flash read 45000000 home`

You can optionally give it a length as well, so I told it to start at "boot" and gave it as long of a length as I could without hitting the end of the flash:

`sunxi_flash read 45000000 boot f80000`

From there I can dump it to an sd card by doing the follow (nice to not have to copy it over the serial connection).

Load the card driver
`sunxi_card0_probe`

Verify you can access the card filesystem
`fatls mmc 0`

Write the image from RAM to the card
`fatwrite mmc 0 45000000 image.bin f80000`

After that I was able to identify the components in the image with binwalk. Extract with dd, and the appropriate subsequent tools (jefferson for JFFS2, unsquash for squash, gzip for kernel).

Now to figure out how to add to it and build a new image.

**Comment by roleoroleo** [comment]
Good!
Normally the fatwrite command is not available.
Yes, probably you can boot directly from SD.
Bu don't worry, if u-boot is ok you can rewrite the flash as you want.
If you share your dump I could help to create a new hack with the first tools: busybox, telnet, ssh...

**Comment by rogerb831** [comment]
Sure, I'd love to get your help with it. It's going to be some work though. It looks like it may be a completely different partition/filesystem structure. Maybe a different SDK. I dunno, I'm a newb.

```
$ binwalk image.bin 

DECIMAL       HEXADECIMAL     DESCRIPTION
--------------------------------------------------------------------------------
0             0x0             uImage header, header size: 64 bytes, header CRC: 0x2057BD0B, created: 2019-12-13 10:11:19, image size: 2080440 bytes, Data Address: 0x40008000, Entry Point: 0x40008000, data CRC: 0xD20F4B45, OS: Linux, CPU: ARM, image type: OS Kernel Image, compression type: none, image name: "ARM OpenWrt Linux-4.9.118"
2268          0x8DC           device tree image (dtb)
10584         0x2958          device tree image (dtb)
14020         0x36C4          device tree image (dtb)
18136         0x46D8          gzip compressed data, maximum compression, from Unix, last modified: 1970-01-01 00:00:00 (null date)
2359296       0x240000        Squashfs filesystem, little endian, version 4.0, compression:gzip, size: 1427049 bytes, 219 inodes, blocksize: 262144 bytes, created: 2019-12-27 01:42:11
3932160       0x3C0000        JFFS2 filesystem, little endian
```



Anyway, not sure I want to post the image up, who knows whats in it. I know my wifi info is for sure but I'm not sure how to strip it out. Email me and I'll send it to you privately. Here a burner email you can find me at fe27eahy53ac@opayq.com.

**Comment by rogerb831** [comment]
Thanks Roleo, I just sent you the details. The burner email is going down. Let me know if it looks like something you can work with or if you think it's a completely new hack effort.

**Comment by roleoroleo** [comment]
Ok, I'm studying the image.
As you can read from binwalk, the image starts with a kernel image in u-boot format.
The first 5 lines that binwalk shows are inside this kernel image.
There is 3 other jffs2 but are empty (only header).
At 0x00f60000 you can find the private info, like serial number, wifi, etc...

Good news, the rootfs contains telnetd.

**Comment by roleoroleo** [comment]
Other questions:

1. The image is less then 16MB. Why do you dump f80000 and not 1000000?
2. u-boot is missing. Is there in the missing part of the flash? Is there an additional internal flash?

**Comment by roleoroleo** [comment]
And after telnetd, you should test dropbear:
https://drive.google.com/open?id=1Y_JKx5qf2BPpVVz4P-AI0tbQ-KTLNEbE

**Comment by rogerb831** [comment]
Nice, I'll give it a shot! 

The reason the image is cut off at the start is because I can't seem to pass a starting RAM address to sunxi_flash, it seems to require a label. So I had to tell it to start at "boot". However that is mapped to a memory address, it's probably after uboot.

I have to use "sunxi_flash read 45000000 **boot** f80000".

I tried "sunxi_flash read 45000000 0x0 f80000", and similar variations, but it just doesn't work.

If you see any indicated or what I could pass instead, or how I can create a new label, I can make another dump.

**Comment by roleoroleo** [comment]
If you run "sunxi_flash 45000000 boot 1000000"?
It works?

**Comment by rogerb831** [comment]
I tried. It gives me a message indicating it passed the end of the FLASH.

On the positive side it looks like it did the update, and the home partition appears normal.
```
---tmp/sd/home_y20gam exist, update begin---
1+0 records in
1+0 records out
22 bytes (22B) copied, 0.001261 seconds, 17.0KB/s
```
Full log at https://www.dropbox.com/s/zcwc2krqq2ftvoh/bootlog2.txt?dl=0

I didn't actually see telnetd running. What changes did you make so I can verify?

**Comment by roleoroleo** [comment]
The update appears to have failed because it does not recognize the version.
```
---tmp/sd/home_y20gam exist, update begin---
1+0 records in
1+0 records out
22 bytes (22B) copied, 0.001261 seconds, 17.0KB/s
check version: newver=¦
¦¦¦¦+¦n, curver=8.2.0.0A_201912270941
sh:
¦¦¦¦+¦n: unknown operand
---same version ? update fail---
```

**Comment by rogerb831** [comment]
Seems the file in /home/homver needs to be incremented?

**Comment by roleoroleo** [comment]
Try to remove the "m" from the name of the file: home_y20ga
We try to upgrade during u-boot and not after linux loads.

**Comment by roleoroleo** [comment]
Other possibility.
Did you see these lines in init.sh?
```
if [ -f "/tmp/sd/Factory/factory_test.sh" ]; then
    /tmp/sd/Factory/config.sh
    exit
fi
```

**Comment by rogerb831** [comment]
home_y20ga seems to have had no effect.

That init file is interesting. That seems like a wide open hook to execute any think on the SD card. Worst cast scenario one could load binary compatible files on the sd card an execute them through that script!

Also, I think we were on the right track earlier:
```
    echo "---/home/home_y20gam exist, update begin---"
        dd if=/home/home_y20gam of=/tmp/newver bs=22 count=1
        newver=$(cat /tmp/newver)
```

I think home wasn't updating because it's looking to pull the homever file from a specific block of the home_y20gam image. If it's not built the exact some way as the factory it's probable that the file won't be at that block..

**Comment by roleoroleo** [comment]
In the MStar platform there are 2 methods to load firmwares.

1. A file without "m" (like home_y203c) containing a jffs2 image
2. A file with "m" (like home_y203cm) containing a 7zipped file with a password and a rsa signature

The first one (the method I use) is loaded by u-boot without checks and is useful for recovery processes.
The second one (used by online update) is loaded by a linux script in the final step of the boot.
The content of this archive overwrites the active file system.

The upgrade we tried didn't work because the file is not signed and doesn't match the yi format.
You can verify this procedure and the format in the file /home/base/tools/extpkg.sh
Unforunately we don't have the private key to sign this file.

**Comment by rogerb831** [comment]
Perhaps is possible to use /tmp/sd/Factory/factory_test.sh to defeat the key check?

1. Load SD card with a /tmp/sd/Factory/factory_test.sh that overwrites /home/base/tools/extpkg.sh
2. Clear SD card and load with home_y203cm

**Comment by roleoroleo** [comment]
Yes, I think it's possible to use factory_test.sh

About the binaries I'm trying some tests.
Helloworld works, complex programs give me a segfault.
I have to investigate.

**Comment by rogerb831** [comment]
We are making progress guys. We've got a hack to root the device without having to use a serial adapter. Roleo is figuring out how to port the app stack over. Fingers crossed.

**Comment by WattsC-90** [comment]
@rogerb831 thats brilliant! I cant believe the trouble it is! Wish they would make it just *a little* easier haha!

well done!

**Comment by Nanook** [comment]
Great work. This is my first camera and this firmware looks amazing. Happy you've found a way in. It wasn't looking great for a while there, so well done ;-).

Just out of interest, Was soldering and flashing it always a possibility still? If so would it be similar to flashing something like an esp8285 where you put it in to pair mode and push the new image over the top? I'm new to this, but I've done a couple of wifi LED controllers this way. Just curious... Thanks.

**Comment by roleoroleo** [comment]
The flash is a spi nor.
You can flash it removing from the pcb and using a simple flash programmer (or ftdi board).
Without removing probably you can't.

**Comment by rogerb831** [comment]
The root kit is available at https://github.com/rogerb831/Yi-Home-Allwinner-Root.

**Comment by WattsC-90** [comment]
@rogerb831 thats great news! A major step towards success!

**Comment by Nanook** [comment]
Confirmed working on 8.2.0.0A_201912270941 - It renamed the Factory folder on the SD Card too

**Comment by rogerb831** [comment]
It renames the Factory folder so it down't keep trying to apply the "fix" on every boot. Thanks for confirming!

**Comment by roleoroleo** [comment]
I can't find a useful toolchain for this platform.
If I can't build a compatible one, all the libc will have to be included in the hack.
A lot of space...

**Comment by Nanook** [comment]
Is there a Discord server we could use? This is a new area for me, but it interests me and I'm happy to put some time in to it.

**Comment by rogerb831** [comment]
https://discord.gg/mWbmFj

**Comment by Nanook** [comment]
Ah, missed it. The only day I've not checked this thread.

**Comment by danielharrelson** [comment]
So, I got the camera rooted.

I'm in via telnet, are we able to flash the firmware with DD or something?

**Comment by rogerb831** [comment]
Theres no release yet. It's still being worked on. Only the root hack is available to enable others to contribute.

**Comment by bytesex** [comment]
> Theres no release yet. It's still being worked on. Only the root hack is available to enable others to contribute.

Hi rogerb,
I wanted to participate in examinating the cam, but failed with the rooting process. Crwated the Issue here:
https://github.com/rogerb831/Yi-Home-Allwinner-Root/issues/1

**Comment by rogerb831** [comment]
A new project has been started for the Allwinner platform (8.2.X firmwares) here https://github.com/roleoroleo/yi-hack-Allwinner. As of right now, it's still not complete and we need more contributors.

**Comment by WattsC-90** [comment]
@rogerb831 good progress or not? you were looking for a toolchain before it moved onto discord.. any luck with that?

**Comment by roleoroleo** [comment]
The toolchain is ok and there is a skeleton of the services.
The hack procedure works correctly.

**Comment by micfogas** [comment]
can you send a fresh invite for the discord server? the old link expired. curious about the status of the toolchain and firmware. you have the toolchain it seems? for now I modified the root script to keep the default services running so it's essentially stock with telnetdd and the replaced extprgm script. can always load services directly off the sd if needed, will have to see what you guys have working. looking forward to making this camera useful.

**Comment by andyb2** [comment]
I also have the firmware 8.2.0.0A_201912270941 and was searching Yi's website for a backup of it. They have a spot for it but the link is broken and the resulting page does show what file was asked for
smarthomecam/familymonitor-y20ga/8.2.0.0A_201912270941home_y25gam . When I asked Yi's support to fix the link she sent me a link for y25 no gam and when I told her I had a 9FUSY31 she gave me a link for a y20 file. Neither worked as expected. I did find out my Y31 is a YI Home Camera 1080P AI+
and after researching the chip inside it is an AI chip. Support at YI still hasnt fixed the link

**Comment by roleoroleo** [comment]
If you want a reliable backup do it yourself.
Connect the serial cable to the pcb and follow the instruction above.

**Comment by rogerb831** [comment]
Getting pretty close gang. Roleo's got an alpha with a functional RTSP stream. I'm testing it with my DVR and it's been solid for 24 hours. Here's a 24 invite to the Discord server. https://discord.gg/k5Fq2F @andyb2. Details on the toolchain at https://github.com/roleoroleo/yi-hack-Allwinner/issues/2.

**Comment by andyb2** [comment]
Great news guys

**Comment by rogerb831** [comment]
A working alpha is released. https://github.com/roleoroleo/yi-hack-Allwinner

**Comment by roleoroleo** [comment]
Issue closed.
For the hack for the new model see here:
https://github.com/roleoroleo/yi-hack-Allwinner

**Comment by WattsC-90** [comment]
@roleoroleo how does someone determine whether they have the normal mstar 9FUS or the Allwinner 9FUS? Thats how i got here because I had a 9FUS that says it is supported!:) thanks for the great work! Very impressed!

**Comment by scruff-01** [comment]
There is a Y and number after 9FUS Like 9FUSY31 with a firmware starting with 8.2.0.0A  That would be Allwinner. A firmware with 4.5.0.0 is probably Mstar. Thats what I suspect

**Comment by WattsC-90** [comment]
@scruff-01 good point, feel stupid asking now haha.

**Comment by labilabiano** [comment]
It works whith BFSU ?, the firmware is the same 8.2.0.0A.

**Comment by scruff-01** [comment]
Ask questions in the issue section at the link roleoroleo posted above

**Comment by pauln** [comment]
I found this thread while looking for a working URL for stock firmware - in case anyone else does likewise, here are some working URLs I've managed to figure out:
[8.2.0.0A_201912270941home_y20gam](http://download.us.xiaoyi.com/yifirmware/smarthomecam/familymonitor-y20ga/8.2.0.0A_201912270941home_y20gam)
[8.2.0.0A_202007211802home_y20gam](http://download.us.xiaoyi.com/yifirmware/smarthomecam/familymonitor-y20ga/8.2.0.0A_202007211802home_y20gam)
[8.2.0.0A_202012041528home_y20gam](http://download.us.xiaoyi.com/yifirmware/smarthomecam/familymonitor-y20ga/8.2.0.0A_202012041528home_y20gam)

Per the file names, these all appear to be for the `y20gam` variant.  The last one listed above is the version which the official firmware download page appears to be trying to link to, and (based on the file name) was released in December 2020, making it the most recent (known) version as of when I'm posting this.

**Comment by micfogas** [comment]
Thanks for the info. I may try to flash the latest firmware just to make sure that roleo's hack works on it. If they patched it somehow or another, which I don't think they care enough to do that given how easy it is to gain root, then we will know. I can still go back via console. The "gam" version is the latest that uses the AllWinner chip I believe. Which is the one I have. If you have a current version of the Yi cam, and you're interested in running the hack on it, roleo has another repo for that one specifically. Considering it works in tandem with the stock firmware and still allows you to use the app and cloud if you want, if you're the type who likes to have full control, you should check it out.

________________________________

**Comment by josechema30** [comment]
hi i have a yi ai+ IFUSY44C23 camera.
MODEL YYS.2016. It has the same problem of the fixed yellow light . I did not manage to make it work , I can not reset it. I can not install the firmwire of what they talk so much. I do not know what else to do. I have used a 16gb memory. no matter what I do the light is always yellow.

**Comment by scruff-01** [comment]
make sure it fat32 and the device is off. If that doesn't work try different sd cards

**Comment by josechema30** [comment]
I have tried with 2 cards of 16 and one of 32gb, I have formatted them many times, I have inserted several firmwires and the camera is only with the yellow light on. I don't know what to do anymore.  thanks

**Comment by scruff-01** [comment]
Are you extracting the zip before putting on the sd card? I used an 8GB sd card for my camera's my other larger sd cards did not work.

**Comment by josechema30** [comment]
> Are you extracting the zip before putting on the sd card? I used an 8GB sd card for my camera's my other larger sd cards did not work.

I think this is the only thing I have to try because I don't have an 8 gb sd card

**Comment by scruff-01** [comment]
I have not updated in a long time but I always did extract the files onto the sd card. Just the contents of the zip file

**Comment by josechema30** [comment]
yes, it worked. But the trick was not who hurt her. The camera alone has stayed with the light on and does nothing. I tried to repair it with the trick or with the installation of the original firmware that I have not found anywhere

**Comment by scruff-01** [comment]
can try this  https://drive.google.com/file/d/1JfP1WyAqjvzpyX06VTqt1rtG7upwzkZA/view?usp=sharing

**Comment by josechema30** [comment]
thanks , i need access

**Comment by scruff-01** [comment]
I granted it

**Comment by josechema30** [comment]
![20220628_142929](https://user-images.githubusercontent.com/108096976/176256650-590f3429-f6a7-47a5-870e-e3a61289e593.jpg)
![capture sd](https://user-images.githubusercontent.com/108096976/176256687-fa468f3f-4573-4046-ac6b-8617fc2c70c2.PNG)

nothing happens =(

**Comment by scruff-01** [comment]
Only the 9 files and the factory folder should be on the sd card. If that doesnt work you probably need to find a better sd card.

**Comment by scruff-01** [comment]
I forget what cam this works with but should be ok  https://drive.google.com/file/d/17vky2YBd0otCKR26M72zP47t98bx8Qvh/view?usp=sharing

**Comment by josechema30** [comment]
well, thanks for trying. But it has not worked.

**Comment by scruff-01** [comment]
Is your camera a 9FUS for sure?  If none of that worked I would buy a small sd card off amazon or something.

**Comment by josechema30** [comment]
my camera is IFUS and i have a 16 gb new card, only for that.

**Comment by scruff-01** [comment]
That could be why nothing works. I thought you had a 9FUS
Check this page https://github.com/roleoroleo/yi-hack-Allwinner/wiki/Remove-the-hack
I think yours is y25ga

**Comment by josechema30** [comment]
thanks. 
I have tried with the 3 files and with these from the list as well:

![Captura yi firmnware](https://user-images.githubusercontent.com/108096976/176273905-d395c244-2a02-4505-897f-ef87dd379729.PNG)

**Comment by josechema30** [comment]
The worst thing is that I didn't do anything weird, I disconnected it just to change the room.

**Comment by josechema30** [comment]
![20220629_141308](https://user-images.githubusercontent.com/108096976/176508051-1fe742e5-a4b7-473e-891b-5e99ffd5546c.jpg)

**Comment by roleoroleo** [comment]
It should be a y21ga.
I can send you a file to try to restore the cam but if it's not a y21ga it will completely brick.

**Comment by roleoroleo** [comment]
Here it is:
[backup_y21ga.gz](https://github.com/roleoroleo/yi-hack-MStar/files/9022914/backup_y21ga.gz)

Follow this guide starting from step 7:
https://github.com/roleoroleo/yi-hack-Allwinner-v2/tree/master/unbrick

**Comment by josechema30** [comment]
thnaks. 
It hasn't worked, I think the camera is dead
